cmd/lifecycler/main.go

```diff
@@ -0,0 +1,319 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	elastic "github.com/olivere/elastic/v7"  // Elasticsearch client
+	"go.uber.org/zap"                        // Logging
+	kingpin "gopkg.in/alecthomas/kingpin.v2" // Command line arg parser.
+
+	// AWS clients and stuff.
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/aws/aws-sdk-go/service/sqs"
+
+	esasg "github.com/mintel/elasticsearch-asg"         // Complex Elasticsearch services
+	"github.com/mintel/elasticsearch-asg/cmd"           // Common logging setup func
+	"github.com/mintel/elasticsearch-asg/pkg/es"        // Elasticsearch client extensions
+	"github.com/mintel/elasticsearch-asg/pkg/lifecycle" // Handle AWS Autoscaling Group lifecycle hook event messages.
+	"github.com/mintel/elasticsearch-asg/pkg/squeues"   // SQS message dispatcher
+)
+
+// Request retry count/timeouts.
+const (
+	awsMaxRetries = 3
+	esRetryInit   = 150 * time.Millisecond
+	esRetryMax    = 1200 * time.Millisecond
+)
+
+// defaultURL is the default Elasticsearch URL.
+const defaultURL = "http://localhost:9200"
+
+// Command line opts
+var (
+	queueURL = kingpin.Arg("queue", "URL of SQS queue receiving lifecycle hook events.").Required().String()
+	esURL    = kingpin.Arg("url", "Elasticsearch URL. Default: "+defaultURL).Default(defaultURL).URL()
+)
+
+// The Elasticsearch API only allows the clearing of all master voting exclusions at once,
+// so only one goroutine should clean up the exclusions once all pending lifecycle events have finished.
+// See: https://www.elastic.co/guide/en/elasticsearch/reference/7.0/voting-config-exclusions.html
+var (
+	// votingLock allows only one call to the Elasticsearch /_cluster/voting_config_exclusions API at once.
+	votingLock sync.Mutex
+
+	// votingCount keeps tracking of how many outstanding lifecycle termination
+	// events are being handled for Elasticsearch master-eligible nodes.
+	// Each goroutine handling the termination of a master-eligible node
+	// increments this this by 1 at the start using the `sync/atomic` package,
+	// and decrements by one after the lifecycle event completes.
+	// If the goroutine is the one that decrements down to 0, it will clear the master
+	// voting exclusions list.
+	votingCount int32
```

Comment:
What happens if the lifecycler dies while handling a lifecycle hook event? When it starts up again, would it just allow another autoscaling event?

Reply:
Or would aws wait on some default timeout for the lifecycle hook?

Reply:
I presume the sqs messages are on the queue still but not visible.

Reply:
The hooks are on a timeout anyway. What lifecycler does is wait until just before the timeout, check if the cluster is good to go, and if not extend the timeout some more.

As for lifecycler dying: https://github.com/mintel/elasticsearch-asg/blob/master/cmd/lifecycler/main.go#L220